pep-0695.rst

```diff
@@ -0,0 +1,1103 @@
+
+    class ClassB[T: ("ForwardDeclaration", bytes)]: ...  # OK
+
+    class ClassC[T: ()]: ...  # Type checker error: two or more types required
+
+    class ClassD[T: (str, )]: ...  # Type checker error: two or more types required
+
+    t1 = (bytes, str)
+    class ClassE[T: t1]: ...  # Type checker error: literal tuple expression required
+
+
+If the specified type is not a tuple expression or the tuple expression includes
+complex expression forms that are not allowed in a type annotation, a type
+checker should generate an error. Quoted forward declarations are allowed.
+
+::
+
+    class ClassF[T: (3, bytes)]: ...  # Type checker error: invalid expression form
+
+
+The specified constrained types must be concrete. An attempt to use a generic
+type should be flagged as an error by a type checker.
+
+::
+
+    class ClassG[T: (list[S], str)]: ...  # Type checker error: generic type
+
+
+
+Generic Type Alias
+------------------
+
+We propose to introduce a new statement for declaring type aliases. Similar
+to ``class`` and ``def`` statements, a ``type`` statement defines a scope
+for type parameters. 
+
+::
+
+    # A non-generic type alias
+    type IntOrStr = int | str
+
+    # A generic type alias
+    type ListOrSet[T] = list[T] | set[T]
+
+
+Type aliases can refer to themselves without the use of quotes.
+
+::
+
+    # A type alias that refers to a forward-declared type
+    type AnimalOrVegetable = "Animal" | "Vegetable"
+
+    # A generic self-referential type alias
+    type RecursiveList[T] = T | list[RecursiveList[T]]
+
+
+The ``type`` keyword is a new soft keyword. It is interpreted as a keyword
+only in this part of the grammar. In all other locations, it is assumed to
+be an identifier name.
+
+Type parameters declared as part of a generic type alias are valid only
+when evaluating the right-hand side of the type alias.
+
+As with ``typing.TypeAlias``, type checkers should restrict the right-hand
+expression to expression forms that are allowed within type annotations.
+The use of more complex expression forms (call expressions, trinary operators,
+arithmetic operators, comparison operators, etc.) should be flagged as an
+error.
+
+Type alias expressions are not allowed to use traditional type variables.
+Type checkers should generate an error in this case.
+
+::
+
+    T = TypeVar("T")
+    type MyList = list[T]  # Type checker error: traditional type variable usage
+
+
+We propose to deprecate the existing ``typing.TypeAlias`` introduced in
+:pep:`613`. The new syntax eliminates its need entirely.
+
+
+Runtime Type Alias Class
+------------------------
+
+At runtime, a ``type`` statement will generate an instance of
+``typing.TypeAliasType``. This class represents the type. Its attributes
+include:
+
+* ``__name__`` is a str representing the name of the type alias
+* ``__parameters__`` is a tuple of TypeVar, TypeVarTuple or ParamSpec objects
+* ``__value__`` is the evaluated value of the type alias
```

Comment:
How does this work with recursive type aliases? What is the value of the type alias while it is being evaluated?